subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy

```diff
@@ -1763,11 +1765,30 @@ All Artifacts:
         _ * domainObjectContext.buildPath >> Path.path(buildPath)
         _ * domainObjectContext.model >> RootScriptDomainObjectContext.INSTANCE
 
-        def publishArtifactNotationParser = NotationParserBuilder.toType(ConfigurablePublishArtifact).toComposite()
-        new DefaultConfiguration(domainObjectContext, confName, configurationsProvider, resolver, listenerManager, metaDataProvider,
-            Factories.constant(resolutionStrategy), TestFiles.fileCollectionFactory(), new TestBuildOperationExecutor(), instantiator,
-            publishArtifactNotationParser, Stub(NotationParser), immutableAttributesFactory, rootComponentMetadataBuilder, Stub(DocumentationRegistry),
-            userCodeApplicationContext, domainObjectContext, projectStateRegistry, Stub(WorkerThreadRegistry), TestUtil.domainObjectCollectionFactory(), calculatedValueContainerFactory)
+        def publishArtifactNotationParser = new PublishArtifactNotationParserFactory(
```

Comment:
Why the change in how `publishArtifactNotationParser` is created?

Reply:
Because I need an instance of `PublishArtifactNotationParserFactory` instead of `Factory<NotationParser<Object, ConfigurablePublishArtifact>>`. Unfortunately, the Gradle DI engine can't locate an instance of `Factory<NotationParser<Object, ConfigurablePublishArtifact>>` so I have to use the exact type `PublishArtifactNotationParserFactory`. I spent some time looking at how our service loader logic worked and toyed with the idea of trying to fix it to support locating and injecting a type matching `Factory<NotationParser<Object, ConfigurablePublishArtifact>>` but it seemed like a rabbit hole I didn't want to go down. As such, it was easier to just pass the concrete type here in the tests.

I agree that your proposal would be better.